### ISSUE TITLE
broadcasting inconsistency?

### ISSUE THREAD

**Issue by kyunghyuncho** [pytorch/pytorch#2973]
when we multiply (or any binary operator) two tensor variables, i've noticed the weird behaviour depending on the shapes of those variables:

(a) when x.size()=(1,10) and y.size()=(10), (x\*y).size()=(1,10) (expected)
(b) when x.size()=(10) and y.size()=(1,10), (x\*y).size()=(1,10) (expected)
(c) when x.size()=(10,1) and y.size()=(10), (x\*y).size()=(10,10) (**unexpected**)
(d) when x.size()=(10) and y.size()=(10,1), (x\*y).size()=(10,10) (**unexpected**)
(e) when x.size()=(10,1) and y.size()=(1,10), (x\*y).size()=(10,10) (expected)

the cases (c) and (d) really easily throw off many (or perhaps only me) from debugging, when such cases happen in a loss function that reduces the output into a single scalar. 

my expectation is that broadcasting should first find a matching axis, and i wonder if the current implementation/behaviour is expected and was designed with a certain goal in mind.

**Comment by killeent** [comment]
@gchanan

**Comment by gchanan** [comment]
Hi @kyunghyuncho: these are expected behavior and consistent with NumPy broadcasting, see: http://pytorch.org/docs/master/notes/broadcasting.html for more information, but basically, the dimensions are matched in rightmost order, there is no inference for a matching axis size.

I'm not sure I completely understand your scalar point about (c) and (d) -- could you provide an example?  But if you are doing reductions, you can control whether the (1) in the reduced dimension is preserved via the `keepdim` parameter, for example in `torch.sum`.

**Comment by kyunghyuncho** [comment]
i meant the following scenario:

loss = torch.mean((x - y) ** 2)

loss is a scalar regardless of whether (x-y).size() is (10,10) or (10,1). this makes debugging pretty much impossible :(

**Comment by kyunghyuncho** [comment]
i wonder if we could at least put some kind of warning or automatic ``squeeze``ing in loss functions. after all, loss functions should work with two input tensors of the same size.

**Comment by apaszke** [comment]
You can always add `assert x.size() == y.size()` in your code if you have your own loss function

**Comment by kyunghyuncho** [comment]
yes, i can. though, many people (including many students in my lab) heavily rely on the loss functions available from pytorch, and having such an assertion expression in pytorch by default would be tremendously helpful.

**Comment by apaszke** [comment]
I think most of them should already have fairly strict requirements for sizes. Do you have any specific ones in mind?

**Comment by gchanan** [comment]
We do also have `torch.utils.backcompat.broadcast_warning.enabled=True` (see the documentation above).  Turning that on won't be limited to loss functions but it could help in debugging.

**Comment by apaszke** [comment]
You can turn it on only periodically (e.g. using a context manager) to limit it to certain scopes you want to debug.

**Comment by kyunghyuncho** [comment]
that's helpful. thanks!
